dartdoc_test/lib/src/reporter.dart

```diff
@@ -0,0 +1,76 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:source_span/source_span.dart';
+import 'package:test/test.dart';
+
+abstract class Reporter {
```

Comment:
Perhaps we should consider:
```suggestion
abstract base class Reporter {
```
That way, we force everybody who wants to make a custom `Reporter` to do: `class MyCustomReporter extends Reporter {....`.

At this point in time this doesn't really change anything. But if in the future we wanted to add another method to `Reporter`, it would be a breaking change! We'd have to bump the major version on `package:dartdoc_test`. Unless we force everybody to use `extends Reporter` instead of `implements Reporter`.

**Example**: If in the future we decide we want to add `reportInvalidSourceFile(String filename, String errorMessage)` method to `Reporter`. If we did:
```dart
abstract base class Reporter {
  void reportSourceFile(String filename, void Function(SourceFileReporter) fn);
  void reportInvalidSourceFile(String filename, String errorMessage);
}
```

Then obviously, it'd be a breaking change, because everybody who made a custom reporter like: `class MyCustomReporter extends Reporter {....` wouldn't have an implementation of `reportInvalidSourceFile`.

But if instead we did:
```dart
abstract base class Reporter {
  void reportSourceFile(String filename, void Function(SourceFileReporter) fn);
  void reportInvalidSourceFile(String filename, String errorMessage) {
    // ignore
  }
}
```

Then it wouldn't be a breaking change. Custom reporters like `class MyCustomReporter extends Reporter {....` will continue to work, the new kind of reports about invalid source files will be ignored by `MyCustomReporter`, but since this is a new kind of message, it's might not be important to it.

Thus, we'd be able to roll out a new kind of field or method, so long as it has a default implementation. Though this only works if people who make custom reporters use `extends Reporter`, and not `implements Reporter`. However, we can always force people to use `extends Reporter` by making it an `abstract base class Reporter`